gnosis-buyback/contracts/Buyback.sol

```diff
@@ -0,0 +1,691 @@
+        uint amount,
+        uint balance
+    );
+
+    event Deposit(        
+        address indexed userAddress,
+        address indexed tokenAddress,
+        uint amount
+    );
+
+    event ModifyAuction(
+        uint auctionIndex,
+        uint amount
+    );
+
+    event ModifyBurnAddress(
+        address burnAddress
+    );
+
+    event ModifyToken(
+        address indexed userAddress,
+        address indexed newToken
+    );
+
+    event ModifyTimeInterval(
+        address indexed userAddress,
+        uint timeInterval
+    );
+
+    event Burn (
+        address indexed tokenAddress,
+        address burnAddress,
+        uint amount
+    );
+    
+    event RemoveAuctionIndex(
+        address indexed userAddress,
+        uint auctionIndex,
+        uint amount
+    );
+
+    event RemoveAuction(
+        address indexed userAddress,
+        uint[] auctionIndexes
+    );
+
+    event PostSellOrder(
+        address indexed userAddress,
+        address indexed buyToken,
+        address indexed sellToken,
+        uint auctionIndex,
+        uint amount,
+        uint newSellerBalance
+    );
+
+    event ModifyTip(
+        address indexed userAddress,
+        uint amount
+    );
+
+    event Tip(
+        address indexed userAddress,
+        address tippedAddress,
+        uint amount
+    );
+
+    event ModifyExternalPoke(
+        address indexed userAddress,
+        bool allowExternalPoke
+    );
+
+    /**
+     * @notice Buyback
+     * @param _dx Address of the dutch exchange
+     */
+    function BuyBack(address _dx) public {
+        require(address(_dx) != address(0));
+        dx = DutchExchange(_dx);
+        dutchXAddress = _dx;
+        owner = msg.sender;
+    }
+       
+    /**
+     * @notice addBuyBack
+     * @param _userAddress Address of the user 
+     * @param _buyToken Address of the buy token
+     * @param _sellToken Address of the sell token
+     * @param _burn Should burn the token after buy back success
+     * @param _auctionIndexes Auction indexes the to participate 
+     * @param _auctionAmounts Auction amounts to fill in each auction index
+     * @param _timeInterval Minimum time passed between buyback execution
+     * @param _allowExternalPoke allow external users to poke 
+    */
+    function addBuyBack(
+        address _userAddress,
+        address _buyToken,
+        address _sellToken, 
+        address _burnAddress, 
+        bool _burn, 
+        uint[] _auctionIndexes, 
+        uint[] _auctionAmounts,
+        uint _timeInterval,
+        bool _allowExternalPoke
+        ) public {
+
+        require(address(_userAddress) != address(0));
+        require(address(_buyToken) != address(0));
+        require(address(_sellToken) != address(0));
+        require(_sellToken != _buyToken);
+        require(_auctionIndexes.length == _auctionAmounts.length);
+        require(buybacks[_userAddress].sellToken == address(0)); // ensure the user address doesn't exist
```

Comment:
Yeah they can modify the auction indexes and amount. Since the buyback would execute in an a  uction index.

Reply:
1. The idea was that the contract can be used for multiple projects. It's also possible to deploy a contract per owner if needed.

2. It's the address of the user that wants to create the buyback config. It's also used as an id, since each address is unique.

3. Yeah, so when executing a postSellOrder the contract checks for that the user has enough balance to execute the `sellOrder` if not reverts the transaction. Right now nothing forces you to add the tokens to satisfy them but then when executing the `postSellOrder` the contract ensures you have enough balance before posting the `sellOrder` to the gnosis contract.